### ISSUE TITLE
subscribe caching collide variables/queries from different operations

### ISSUE THREAD

**Issue by 524c** [HoudiniGraphql/houdini#731]
### Describe the bug

Two different operations interfere with each other and doing a store reset after a comparison of changes in variables, which did not occur.

### Severity

serious, but I can work around it

### Steps to Reproduce the Bug

I'm working on a mini project and I'll update this report soon. For now I made two videos to demonstrate what happens.
After an initial investigation it seems to me that the problem is in the logic of the cache.

The videos below are using the current version 0.17.13 and a simple Svelte project page.

Each button does a different query. The expected behavior is that the result would be persisted in the store, but one operation interferes with the other.
Fisrt video without any changes in Houdini code:

https://user-images.githubusercontent.com/527641/204096718-beffa950-008d-42d0-bd72-a1ba878d1eec.mp4

This second video is after commenting the refreshSubscription function. That way I managed to keep the data of both operations.

Note: in this quick test I made direct changes to the volatile runtime files.
$houdini/plugins/houdini-svelte/runtime/stores/query.js
```
  refreshSubscription(newVariables) {
    /*
    const cache = getCache();
    if (this.subscriptionSpec) {
      cache.unsubscribe(this.subscriptionSpec, this.lastVariables || {});
    }
    this.subscriptionSpec = {
      rootType: this.artifact.rootType,
      selection: this.artifact.selection,
      variables: () => newVariables,
      set: (newValue) => this.store.update((s) => ({ ...s, data: newValue }))
    };
    cache.subscribe(this.subscriptionSpec, newVariables);
    this.lastVariables = newVariables;*/
  }
```
Result after this "code change":

https://user-images.githubusercontent.com/527641/204097169-08e1d188-492d-4abc-b68d-c7fa8b47c285.mp4

+page.svelte
```
<script lang="ts">
  import AllUsers from '$lib/components/utils/AllUsers.svelte';
  import UserPk from '$lib/components/utils/UserPk.svelte';

  export let data: any;

  $: ({ userPk } = data);
</script>

<br />

<AllUsers />
<br />

<UserPk pk={1} />
<br />
```

Component UserPk.svelte
```
<script lang="ts">
  import { G_userPk } from '$houdini';
  import { Button } from 'flowbite-svelte';

  export let pk = 0;
  const startColor = 'mt-2 flex flex-wrap text-red-500';
  const normalColor = 'mt-2 flex flex-wrap text-blue-500';
  $: color = startColor;

  export function userPkVariables({ props }: { props: { pk?: number } }) {
    return {
      pk: props.pk
    };
  }

  async function userPk() {
    color = startColor;
    await G_userPk.fetch({ variables: { pk } });
    setTimeout(() => {
      color = normalColor;
    }, 1000);
  }
</script>

<Button on:click={userPk}>userPk</Button>
<br />

{#if $G_userPk?.data}
  <pre class={color}>G_userPk: {JSON.stringify($G_userPk?.data, null, 2)}</pre>
  <br />
{/if}
```

Component AllUsers.svelte
```
<script lang="ts">
  import { G_allUsers } from '$houdini';
  import { Button } from 'flowbite-svelte';
  const startColor = 'mt-2 flex flex-wrap text-red-500';
  const normalColor = 'mt-2 flex flex-wrap text-blue-500';

  $: color = startColor;

  async function allUsers() {
    color = startColor;
    await G_allUsers.fetch();
    setTimeout(() => {
      color = normalColor;
    }, 1000);
  }
</script>

<Button on:click={allUsers}>allUsers</Button>
<br />

{#if $G_allUsers?.data}
  <pre class={color}>G_allUsers: {JSON.stringify($G_allUsers?.data, null, 2)}</pre>
  <br />
{/if}
```

**Comment by AlecAivazis** [comment]
Thanks for reporting this! 

I think I am missing what the issue is - it looks like there are 2 stores with different values, is that not what you want? Could you share the actual queries powering the stores?

**Comment by 524c** [comment]
> Thanks for reporting this!
> 
> I think I am missing what the issue is - it looks like there are 2 stores with different values, is that not what you want? Could you share the actual queries powering the stores?

Hello Alec,

I made an update and included the two queries.

This is a generic example I took from an ongoing project. These two components do different operations, but one interferes with the content of the other. Am I doing something wrong on Svelte's side?

**Comment by 524c** [comment]
$houdini/plugins/houdini-svelte/runtime/stores/query.js
```
 async fetch(args) {
    const config = await this.getConfig();
    const { policy, params, context } = await fetchParams(this.artifact, this.storeName, args);
    const isLoadFetch = Boolean('event' in params && params.event);
    const isComponentFetch = !isLoadFetch;
    const input =
      (await marshalInputs({
        artifact: this.artifact,
        input: params?.variables
      })) || {};
    const newVariables = {
      ...this.lastVariables,
      ...input
    };
    console.log(`this.lastVariables: ${this.lastVariables}\nnewVariables: ${newVariables}`);

    let variableChange = !deepEquals(this.lastVariables, newVariables);
    if (variableChange && isBrowser) {
      this.refreshSubscription(newVariables);
      this.store.update((s) => ({ ...s, variables: newVariables }));
    }
    if (this.loadPending && isComponentFetch) {
      log.error(`\u26A0\uFE0F Encountered fetch from your component while ${this.storeName}.load was running.
This will result in duplicate queries. If you are trying to ensure there is always a good value, please a CachePolicy instead.`);
      return get(this.store);
    }

```

I added a log and the store update occurs comparing changes between variables from two different queries. But actually only the userPk query uses a variable. The allUsers query doesn't even use a variable.
<img width="690" alt="console log" src="https://user-images.githubusercontent.com/527641/204099301-ed7d4b25-2856-453e-beed-d1fe38786070.png">

These logs are related to the first video, when I alternately clicked on the buttons that make different queries.

**Comment by AlecAivazis** [comment]
Ah so the problem is not in the underlying cache but that the store state resets?

**Comment by 524c** [comment]
> Ah so the problem is not in the underlying cache but that the store state resets?

Yes, that's it. The problem I see is in the reset of the store. When I started investigating I thought it was something with the cache.

**Comment by jycouet** [comment]
I created a branch here [issue/twoOperations](https://github.com/HoudiniGraphql/houdini/tree/issue/twoOperations)
And in `/e2e/sveltekit/src/routes/twoOperations` you will see your components & queries.

But was not able to reproduce.

Maybe you could have a look @524c and let us know what's the delta?

To get it:
```bash
# 1/ clone the repo

# then
pnpm i
pnpm build
pnpm --filter sveltekit dev

# check there
# http://localhost:5173/twoOperations
```

**Comment by 524c** [comment]
Hello @jycouet, I'm preparing a mini project to show how to reproduce and I'm going up on stackblitz.

**Comment by jycouet** [comment]
Maybe you can start with the branch I just created with a fork?

**Comment by 524c** [comment]
> Maybe you can start with the branch I just created with a fork?

Yes, I will. I noticed that the problem only occurs when my graphql server responds with a fragment. I'll soon do a PR on this branch. Thanks.

**Comment by jycouet** [comment]
Just before doing all this, do you have a `+page.ts` and what's inside (if you have one)?

**Comment by 524c** [comment]
> Just before doing all this, do you have a `+page.ts` and what's inside (if you have one)?

In this example I made the report I don't have a +page.ts.

When I noticed this problem, I had, initially, a +page.ts to bring a content already loaded from the server.

**Comment by 524c** [comment]
@jycouet, bringing more details. I'm doing queries using fragments, with responses of different types. This link has a great example => https://graphql.org/learn/queries/#inline-fragments

I'll be back shortly..

**Comment by 524c** [comment]
Hey @jycouet

In my project I use [Pothos](https://pothos-graphql.dev/) to build schema.

And I especially use the [pothos validation plugin + zod](https://pothos-graphql.dev/docs/plugins/validation) in my resolvers and [pothos errors plugin](https://pothos-graphql.dev/docs/plugins/errors) which generate types at runtime for different responses. So in case of success it responds with one type and in case of error with another.

I tried to emulate the pothos plugins I have on my system without making big changes to the [branch twoOperations](https://github.com/HoudiniGraphql/houdini/tree/issue/twoOperations). I wrote the schemas in `houdini/e2e/_api/schema.graphql` to respond the same way my whole system does, but it didn't work because the types are generated at runtime by Pothos, which is what I have on my system, and that's a bit complex to "emulate" without Pothos 😂

I will need to set up a specific Graphql server (Yoga) that responds with these types/union, where I query using fragments, which is where this store reset problem happens.

There's probably still a simpler way to do this, but I know it would take longer trying to figure it out. My current minimum stack to be able to generate something reproducible would be this:
* Houdini
* Yoga Graphql
* Pothos + Error and Validate plugins

If you can take a look at this code I'll share the repository with you as soon as I have it ready. I'm going to take some time now to prepare all this 😅

**Comment by 524c** [comment]
Hey @jycouet and @AlecAivazis 

In my project I use [Pothos](https://pothos-graphql.dev/) to build schema.

And I especially use the [pothos validation plugin + zod](https://pothos-graphql.dev/docs/plugins/validation) in my resolvers and [pothos errors plugin](https://pothos-graphql.dev/docs/plugins/errors) which generate types at runtime for different responses. So in case of success it responds with one type and in case of error with another.

I tried to emulate the pothos plugins I have on my system without making big changes to the [branch twoOperations](https://github.com/HoudiniGraphql/houdini/tree/issue/twoOperations). I wrote the schemas in `houdini/e2e/_api/schema.graphql` to respond the same way my whole system does, but it didn't work because the types are generated at runtime by Pothos, which is what I have on my system, and that's a bit complex to "emulate", which is hard to do without it 😂

I will need to set up a specific Graphql server (Yoga) that responds with these types/union, where I query using fragments, which is where this store reset problem happens.

There's probably still a simpler way to do this, but I know it would take longer trying to figure it out. My current minimum stack to be able to generate something reproducible would be this:
* Houdini
* Yoga Graphql
* Pothos + Error and Validate plugins

If you can take a look at this code I'll share the repository with you as soon as I have it ready. I'm going to take some time now to prepare all this 😅

**Comment by jycouet** [comment]
Thx for the context.
If it's easier, you don't have to do it in the Houdini e2e test project.

Having a minimal repro with your stack would be intresting also 👍

**Comment by 524c** [comment]
> Thx for the context. If it's easier, you don't have to do it in the Houdini e2e test project.
> 
> Having a minimal repro with your stack would be intresting also 👍

Perfect, I was already redoing all e2e resolvers to keep them compatible with Pohos lol
Okay, I'll generate a minimal version then and share it as soon as possible. Thanks 👍

**Comment by 524c** [comment]
> ### Describe the bug
> Two different operations using, inline fragments, interfere with each other and doing a store reset.
> 
> In my project I use [Pothos](https://pothos-graphql.dev/) to build schema.
> 
> And I especially use the [pothos validation plugin + zod](https://pothos-graphql.dev/docs/plugins/validation) in my resolvers and [pothos errors plugin](https://pothos-graphql.dev/docs/plugins/errors) which generate types at runtime for different responses. So in case of success it responds with one type and in case of error with another.
> 
> ### Severity
> blocking all usage of Houdini
> 
> ### Steps to Reproduce the Bug
> 1- clone the repo
> 
> ```
> git clone https://github.com/524c/houdini-issue-twoOperations
> ```
> 
> then
> 
> ```
> cd houdini-issue-twoOperations
> pnpm i
> pnpm dev
> ```
> 
> check there
> 
> http://localhost:3000/twoOperations
> 
> ### Reproduction
> The videos below are using the current version 0.17.13 and a simple Svelte project page.
> 
> Each button does a different query. The expected behavior is that the result would be persisted in the store, but one operation interferes with the other. Fisrt video without any changes in Houdini code:
> 
>  houdini-svelte-store-bug.mp4 
> This second video is after commenting the refreshSubscription function. That way I managed to keep the data of both operations.
> 
> Note: in this quick test I made direct changes to the volatile runtime files. $houdini/plugins/houdini-svelte/runtime/stores/query.js
> 
> ```
>   refreshSubscription(newVariables) {
>     /*
>     const cache = getCache();
>     if (this.subscriptionSpec) {
>       cache.unsubscribe(this.subscriptionSpec, this.lastVariables || {});
>     }
>     this.subscriptionSpec = {
>       rootType: this.artifact.rootType,
>       selection: this.artifact.selection,
>       variables: () => newVariables,
>       set: (newValue) => this.store.update((s) => ({ ...s, data: newValue }))
>     };
>     cache.subscribe(this.subscriptionSpec, newVariables);
>     this.lastVariables = newVariables;*/
>   }
> ```
> 
> Result after this "code change":
> 
>  houdini-svelte-store-with.code.changes.mp4

Hey @jycouet and @AlecAivazis, I just edited the report and included a link to a repro for this bug.

**Comment by jycouet** [comment]
Thx for the repo with the repro 😉
It will be easier to look what is going on there as I was able to reproduce it locally 👍

Looking quickly at the schema & houdini's artifacts, I think that it's link to the return union with `... on ZodError`, `... on Error`.
I will have a deeper look tomorrow.

Thx🙏

**Comment by 524c** [comment]
> 

Thanks. If you need help or have any questions, call me 😉

**Comment by 524c** [comment]
> Thx for the repo with the repro 😉
> It will be easier to look what is going on there as I was able to reproduce it locally 👍
> 
> Looking quickly at the schema & houdini's artifacts, I think that it's link to the return union with `... on ZodError`, `... on Error`.
> I will have a deeper look tomorrow.
> 
> Thx🙏

@jycouet, I made small fixes in the repository 👍

**Comment by AlecAivazis** [comment]
Okayyy finally got some time to pull down the reproduction and figure out what was going on. Thanks so much for putting that together. 

The quick answer is that the unions in the query results are confusing the cache when its trying to read values. Message is not nullable so when its reading the values back from the cache, it sees a missing `message` and cascades the `null` value all the way up, destroying the previous value. This will be quite an involved fix but I'll try to get it working in the next few days. Luckily, the fix I have in mind is also going to address #651 which makes me happy :D. In the meantime, you can get things working better by removing the `!` in the definition for the message field. It's not a perfect fix since if the error state changes between requests, you'll just run into the problem again but at least it's something.

Thanks for being patient!

**Comment by 524c** [comment]
Hi @AlecAivazis, I thank you and @jycouet for this project. I also didn't have time to look deeper into this bug, but whenever I can I want to help improve Houdini 🚀🚀

**Comment by AlecAivazis** [comment]
Just an update - #747 should fix this. We reproduced in our e2e test suite and everything is working 🎉

**Comment by 524c** [comment]
> Just an update - #747 should fix this. We reproduced in our e2e test suite and everything is working 🎉

Thank you very much @AlecAivazis and @jycouet 😁
